Typed Settings

NobodysNightmare opened the pull request:
## Work Package

I see this PR as part of general housekeeping tasks around OpenProject: https://community.openproject.org/work_packages/18714
## Description

Calling `Setting.foo_setting` currently most of the times returns a string in many cases where you actually expect other types to be returned (such as numbers and booleans).

This PR proposes to use the already existing and sporadically used `format` option of the settings to correctly type the return value of `Setting` calls.

I can see this breaking in many places, because currently code is written like this:

```
if Setting.foo_setting == '1'
  call_me Setting.bar_setting.to_i
end
```

But I think I should be worth it long-term-ish. Just want to gather some feedback before adding format information to the whole `settings.yml`.
## Considerations
- stuff will break
- you need to clear the rails cache for changes in return values to become visible

NobodysNightmare committed:
allow more typed results from settings

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-typed-settings.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-typed-settings.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

ulferts commented:
> Whatever it is, I fear the Greeks even when they bring gifts.

I like the approach but am unsure about the backlash. Would be good for configuration.rb, too.

linki commented:
:+1: for typed settings (and configuration). the implementation is terrible. (no offense, it fits the rest of the code.) i would seriously consider refactoring that part at some point as settings are a very central aspect. therefore, i strongly suggest you to write some tests for this.

NobodysNightmare commented:
@linki Any idea on how to do proper tests here?

I don't really want to use existing settings as source for my mock data (e.g. "an int setting should be returned as integer). But since there is no good DI possibility present this seems to be my only choice (apart from doing a larger refactoring of settings in general that would allow for better DI)

NobodysNightmare committed:
make settings helper robust against missing dup

seems like dup will not work for certain types, but for those it will still be implemented, but raise an exception.

NobodysNightmare committed:
remove to_i wherever a format was specified

- This should be safe since we care about the format specified in settings
- Lets see if anything breaks

NobodysNightmare commented on spec/models/user_spec.rb:
```diff
@@ -329,7 +329,7 @@ def denied_auth_source.allow_password_changes?; false; end
 
     it { expect(@u.valid?).to be_falsey }
-    it { expect(@u.errors[:password]).to include I18n.t('activerecord.errors.messages.too_short', count: Setting.password_min_length.to_i) }
+    it { expect(@u.errors[:password]).to include I18n.t('activerecord.errors.messages.too_short', count: Setting.password_min_length) }
```
Shut up hound, I already shortened it!

ulferts commented:
Idle musings: What about the plugins?

linki commented on spec/support/settings.rb:
```diff
@@ -31,9 +31,14 @@
+      # try to store a duplicate, but use the original where :dup is not available (int, bool, sym)
+      # apparently this is the only way to determine if dup will work
+      h[k] = Setting[k].dup
+    rescue
```
you can at least define the `dup`-related exception name here

NobodysNightmare replied:
How do I know that is the only one for undupable classes?

linki replied:
you don't

linki commented:
@NobodysNightmare i'm fine with taking some real, stable, obviously typed settings to test the formatting stuff. e.g. app_title for strings, port for integer, some default if `format` is not defined.

just make sure if someone changes the format of one of your setting-under-test she knows quickly what the reason for the failing test is.

NobodysNightmare commented:
I won't have time for this soon-ish, still I think it is a good idea.

Will come back if I get it to a more finished state.

(to the ones able to do so: **don't delete my branch** ;-))
